### ISSUE TITLE
Beacon Node Crashes After Large Slot Skip

### ISSUE THREAD

**Issue by djrtwo** [prysmaticlabs/prysm#622]
# Issue
After running `beacon-chain` and `validator` for ~800 slots, it appears as though my system went to sleep for 40 minutes. When it came back online ~300 slots were skipped (as expected), but then the `beacon-chain` process threw an uncaught error: `runtime error: slice bounds out of range`. 

At this point the `validator` started logging errors about not receiving anything from the RPC stream (as expected).

Relevant beacon dump:
```
[2018-10-03 16:10:55]  INFO blockchain: New beacon slot slotNumber=802
[2018-10-03 16:10:55]  INFO blockchain: Applying fork choice rule
[2018-10-03 16:10:55]  INFO blockchain: Canonical block determined blockHash=0x681ae185f88329fb44b5412f82f874b8fda5e28f04f1bdbd4d16d9b409da3ed6
[2018-10-03 16:10:55]  INFO blockchain: Finished processing received block: 0x73fb30416bcd41c75905ceae285ee6baac20dfc030b022adadd61de92e452872
[2018-10-03 16:52:05]  INFO blockchain: New beacon slot slotNumber=1111
[2018-10-03 16:52:05]  INFO blockchain: Applying fork choice rule
[2018-10-03 16:52:05]  INFO blockchain: Canonical block determined blockHash=0x73fb30416bcd41c75905ceae285ee6baac20dfc030b022adadd61de92e452872
[2018-10-03 16:52:05]  INFO blockchain: Finished processing received block: 0x6d3b9499c2534c6382347a454e51b76231562afb5a9123d529513a470d7e8615
[2018-10-03 16:52:12]  INFO blockchain: New beacon slot slotNumber=1112
[2018-10-03 16:52:12]  INFO blockchain: Applying fork choice rule
panic: runtime error: slice bounds out of range

goroutine 86 [running]:
github.com/prysmaticlabs/prysm/beacon-chain/types.(*ActiveState).calculateNewBlockHashes(0xc000323b30, 0xc00000eb10, 0x322, 0x2, 0xc00000eae8, 0x1, 0x1, 0x1)
    beacon-chain/types/active_state.go:140 +0x1c5
github.com/prysmaticlabs/prysm/beacon-chain/types.(*ActiveState).CalculateNewActiveState(0xc000323b30, 0xc00000eb10, 0xc0002dc888, 0x322, 0x7228452ee91dd600, 0xc00000eb20, 0x0, 0x0)
    beacon-chain/types/active_state.go:214 +0x106
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*ChainService).updateHead(0xc0001641e0, 0xc000093f80)
    beacon-chain/blockchain/service.go:289 +0x5d3
created by github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*ChainService).Start
    beacon-chain/blockchain/service.go:106 +0x1fe
```

Relevant validator dump
```

[2018-10-03 16:10:47]  INFO beacon: role: PROPOSER, assigned slot: 797, current slot: 801
[2018-10-03 16:10:47]  INFO beacon: Validator with pub key 0xCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC re-assigned to shard ID 0 for PROPOSER duty at slot 802
[2018-10-03 16:10:55]  INFO beacon: role: PROPOSER, assigned slot: 802, current slot: 802
[2018-10-03 16:10:55]  INFO beacon: Assigned proposal slot number reached slotNumber=802
[2018-10-03 16:10:55]  INFO proposer: Performing proposer responsibility
[2018-10-03 16:10:55]  INFO proposer: Block proposed successfully with hash 0x681ae185f88329fb44b5412f82f874b8fda5e28f04f1bdbd4d16d9b409da3ed6
[2018-10-03 16:52:05]  INFO beacon: role: PROPOSER, assigned slot: 802, current slot: 1111
[2018-10-03 16:52:12]  INFO beacon: role: PROPOSER, assigned slot: 802, current slot: 1112
[2018-10-03 16:52:12] ERROR beacon: Could not receive latest validator assignment from stream: rpc error: code = Unavailable desc = transport is closing
[2018-10-03 16:52:12] ERROR beacon: Could not receive latest validator assignment from stream: rpc error: code = Unavailable desc = transport is closing
[2018-10-03 16:52:12] ERROR beacon: Could not receive latest attestation from stream: rpc error: code = Unavailable desc = transport is closing
[2018-10-03 16:52:12] ERROR beacon: Could not receive latest attestation from stream: rpc error: code = Unavailable desc = transport is closing
[2018-10-03 16:52:12] ERROR beacon: Could not receive latest attestation from stream: rpc error: code = Unavailable desc = transport is closing
```

Note: the validator continued to run with similar output for another couple of hours before I came back to the computer.

# Potential cause of error
There is a known bug in the spec regarding how `active_state.recent_block_hashes` is updated and used after long skips in slots between parent/child. I plan on proposing a fix to this issue tomorrow and will link to it here.

**Comment by djrtwo** [comment]
The related fix in the spec is in this PR https://github.com/ethereum/eth2.0-specs/pull/44 (still under review)

**Comment by rauljordan** [comment]
This is also affected by the RPC connection idling for a while or disconnecting. Our current validator client does not handle gRPC disconnections gracefully. Instead, it should try to reconnect and run its main loop from scratch in the case of an error.

**Comment by djrtwo** [comment]
Right, one the validator side for sure.

The particulars of the runtime error in the Beacon node make me strongly believe that it was something related to the large block skip setting off this error in calculating the new `recent_block_hashes`.

It's a `runtime error: slice bounds out of range` when running `calculateNewBlockHashes` inside of `CalculateNewActiveState`.

**Comment by rauljordan** [comment]
@rawfalafel @terenc3t can you guys confirm your latest fixes resolve this issue?

**Comment by terenc3t** [comment]
I think this problem is primarily due to initial sync is not working. If the node disconnects for more than CYCLE_LENGTH, it'll fail

**Comment by rauljordan** [comment]
Closed by #743
